### DISCUSSION TITLE
How to avoid fatal mistakes when modifying the agent

### DISCUSSION THREAD

**Discussion by huange7** [open-telemetry/opentelemetry-java-instrumentation#12026] [Q&A] [answered]
Hello, I encountered some problems after modifying the javaagent.

In order to make the Instrumenter of the agent not work in certain situations and be adjustable dynamically, we added some code to the Instrumenter, such as: 
```
private Context doStart(Context parentContext, REQUEST request, @Nullable Instant startTime) {
    if (Util.shouldInterrupt(instrumentationName)) {
      return parentContext;
    }
    ...
}
```
But when I make this modification, if the Instrumenter returns directly, this will cause some keys in the context to have null values, which may cause a NullPointerException.
like here: https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation/undertow-1.4/bootstrap/src/main/java/io/opentelemetry/javaagent/bootstrap/undertow/UndertowActiveHandlers.java#L36
```
 public static void increment(Context context) {
    context.get(CONTEXT_KEY).incrementAndGet();
  }
```

Here are my problems:
* How many such usages(get value in the context before Instrumenter.start) are there in the agent?
* Is there any means of detection?
* Is this operation of the instrumenter returning directly allowed?

**Comment by trask** [comment, marked as answer]
have you seen [`SpanSuppressor.java`](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/instrumenter/SpanSuppressor.java), does that work for your use case?

**Reply by huange7** [reply]
Thanks for your reply! It seems that SpanSuppressor does not provide a way for custom implementation. If I want to add a new type, do I have to add it here:
```
static SpanSuppressionStrategy fromConfig(@Nullable String value) {
    if (value == null) {
      value = "semconv";
    }
    switch (value.toLowerCase(Locale.ROOT)) {
      case "none":
        return NONE;
      case "span-kind":
        return SPAN_KIND;
      case "my-case":
        return MY_CASE;
      default:
        return SEMCONV;
    }
  }
```
